src/neo/Ledger/Blockchain.cs

```diff
@@ -197,6 +197,9 @@ public Block GetBlock(UInt256 hash)
 
         public UInt256 GetBlockHash(uint index)
         {
+            var kvp = block_cache.FirstOrDefault(kvp => kvp.Value.Index == index);
+            if (kvp.Key != null)
+                return kvp.Value.Hash;
```

Comment:
Why?

Reply:
We should get the hash from block_cache first. Otherwise you cannot get block in block_cache by Index.

Reply:
Why? The hash in `block_cache` must also be in `header_index`. Because the `block_cache` stores all the blocks that have been persisted.

Reply:
>  the block_cache stores all the blocks that have been persisted.

No, `block_cache` will also store the block to be persisted, which is not contained in `header_index`.

https://github.com/neo-project/neo/blob/a995e029f3b59b0d350181e855390009ed240c55/src/neo/Ledger/Blockchain.cs#L333-L335

Reply:
`Dictionary` is not thread-safe, you may modify it while traversing the `block_cache`. I think we should delete this paragraph directly.